File: scripts/lib/pre_broker_demand.mjs
import { createHash } from "node:crypto";

export const CANONICAL_PRE_BROKER_DEMAND_VERSION =
  "canonical-pre-broker-demand/1.0";

const SUPPORTED_SOURCE_VERSIONS = new Set([
  "pre-broker-model-demand/1.0",
  "pre-broker-model-demand/2.0",
]);

function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.keys(value).sort().map((key) => [key, canonical(value[key])]),
    );
  }
  return value;
}

function digest(value) {
  return createHash("sha256")
    .update(`${JSON.stringify(canonical(value))}\n`)
    .digest("hex");
}

/** Normalize both public demand contracts to one runtime-owned node shape. */
export function normalizePreBrokerDemand(graph) {
  if (!graph || typeof graph !== "object") {
    throw new Error("Pre-broker demand graph is absent.");
  }
  if (!SUPPORTED_SOURCE_VERSIONS.has(graph.schema_version)) {
    throw new Error(`Unsupported pre-broker demand version ${graph.schema_version ?? "<missing>"}.`);
  }
  const body = Object.fromEntries(
    Object.entries(graph).filter(([key]) => key !== "graph_sha256"),
  );
  if (graph.graph_sha256 !== digest(body)) {
    throw new Error("Pre-broker demand has a stale canonical graph hash.");
  }
  const forecastPeriods = [...(graph.forecast_periods ?? [])];
  if (forecastPeriods.length !== 3) {
    throw new Error("Pre-broker demand requires exactly three forecast periods.");
  }
  const v2 = graph.schema_version === "pre-broker-model-demand/2.0";
  const nodes = (graph.nodes ?? []).map((node) => {
    const sourceLineId = node.source_line_id ?? null;
    const allowedAuthorities = [...(node.allowed_authorities ?? [])];
    return {
      node_id: String(node.node_id ?? ""),
      node_kind: v2 ? String(node.node_kind ?? "model_demand") : "filed_observation",
      section: String(node.section ?? ""),
      source_line_ids: sourceLineId ? [String(sourceLineId)] : [],
      metric_id: v2 && node.metric_id ? String(node.metric_id) : null,
      label: String(node.label ?? ""),
      period_end: String(node.period_end ?? ""),
      material: node.material !== false,
      broker_demand_eligible: v2
        ? node.broker_demand_eligible === true
        : allowedAuthorities.includes("selected_broker"),
      allowed_authorities: allowedAuthorities,
      definition_signature_sha256:
        String(node.definition_signature_sha256 ?? ""),
      consumer_ids: v2
        ? [...(node.consumer_ids ?? [])].map(String).sort()
        : [],
    };
  }).sort((left, right) => left.node_id.localeCompare(right.node_id));
  return Object.freeze({
    schema_version: CANONICAL_PRE_BROKER_DEMAND_VERSION,
    source_schema_version: graph.schema_version,
    source_graph_sha256: graph.graph_sha256,
    run_id: String(graph.run_id ?? ""),
    forecast_periods: forecastPeriods,
    reporting_currency: String(graph.reporting_currency ?? ""),
    units: String(graph.units ?? ""),
    nodes,
  });
}

File: scripts/pre_broker_demand.py
#!/usr/bin/env python3
"""Version-aware pre-broker demand reader with one canonical node form."""

from __future__ import annotations

import hashlib
import json
from typing import Any


CANONICAL_PRE_BROKER_DEMAND_VERSION = "canonical-pre-broker-demand/1.0"
SUPPORTED_SOURCE_VERSIONS = {
    "pre-broker-model-demand/1.0",
    "pre-broker-model-demand/2.0",
}

LEGACY_METRIC_ALIASES = {
    "revenue": "revenue", "revenues": "revenue", "turnover": "revenue",
    "ebit": "ebit", "adjusted ebitda": "adjusted_ebitda", "ebitda": "adjusted_ebitda",
    "depreciation and amortisation": "depreciation_and_amortisation",
    "depreciation and amortization": "depreciation_and_amortisation",
    "effective tax rate": "effective_tax_rate", "capex": "capex",
    "capital expenditure": "capex", "change in working capital": "change_in_working_capital",
    "dividends": "dividends", "dividends paid": "dividends",
    "share buybacks": "share_buybacks", "share repurchases": "share_buybacks",
}


def _legacy_metric_id(node: dict[str, Any]) -> str | None:
    label = " ".join(str(node.get("label") or "").lower().replace("_", " ").split())
    if label in LEGACY_METRIC_ALIASES:
        return LEGACY_METRIC_ALIASES[label]
    prefix = str(node.get("node_id") or "").split(".", 1)[0]
    return LEGACY_METRIC_ALIASES.get(prefix.replace("_", " "))


def canonical_hash(value: Any) -> str:
    encoded = (
        json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
        + "\n"
    ).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def normalize_pre_broker_demand(graph: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(graph, dict):
        raise ValueError("Pre-broker demand graph is absent.")
    source_version = graph.get("schema_version")
    if source_version not in SUPPORTED_SOURCE_VERSIONS:
        raise ValueError(f"Unsupported pre-broker demand version {source_version or '<missing>'}.")
    body = {key: value for key, value in graph.items() if key != "graph_sha256"}
    if graph.get("graph_sha256") != canonical_hash(body):
        raise ValueError("Pre-broker demand has a stale canonical graph hash.")
    forecast_periods = list(graph.get("forecast_periods") or [])
    if len(forecast_periods) != 3:
        raise ValueError("Pre-broker demand requires exactly three forecast periods.")
    v2 = source_version == "pre-broker-model-demand/2.0"
    nodes: list[dict[str, Any]] = []
    for node in graph.get("nodes") or []:
        source_line_id = node.get("source_line_id")
        allowed_authorities = list(node.get("allowed_authorities") or [])
        nodes.append({
            "node_id": str(node.get("node_id") or ""),
            "node_kind": str(node.get("node_kind") or "model_demand") if v2 else "model_demand",
            "section": str(node.get("section") or ""),
            "source_line_ids": [str(source_line_id)] if source_line_id else [],
            "metric_id": str(node.get("metric_id")) if v2 and node.get("metric_id") else _legacy_metric_id(node),
            "label": str(node.get("label") or ""),
            "period_end": str(node.get("period_end") or ""),
            "material": node.get("material") is not False,
            "broker_demand_eligible": (
                node.get("broker_demand_eligible") is True
                if v2
                else "selected_broker" in allowed_authorities
            ),
            "allowed_authorities": allowed_authorities,
            "definition_signature_sha256": str(node.get("definition_signature_sha256") or ""),
            "consumer_ids": sorted(str(value) for value in (node.get("consumer_ids") or [])) if v2 else [str(node.get("node_id") or "")],
        })
    nodes.sort(key=lambda item: item["node_id"])
    return {
        "schema_version": CANONICAL_PRE_BROKER_DEMAND_VERSION,
        "source_schema_version": source_version,
        "effective_schema_version": "pre-broker-model-demand/2.0",
        "migration_status": "native_v2" if v2 else "migrated_v1_to_v2",
        "source_graph_sha256": graph["graph_sha256"],
        "run_id": str(graph.get("run_id") or ""),
        "forecast_periods": forecast_periods,
        "reporting_currency": str(graph.get("reporting_currency") or ""),
        "units": str(graph.get("units") or ""),
        "nodes": nodes,
    }
